### DISCUSSION TITLE
async enumerable should not error if it throws without yield statement

### DISCUSSION THREAD

**Discussion by JeffreyMercado** [dotnet/csharplang#3145] [Language Ideas] [answered]
I'm stubbing out an application making all methods throw a `NotImplementedException`. But it seems I cannot do this for an async generator, I will get a CS8420 error.

> error CS8420: The body of an async-iterator method must contain a 'yield' statement. Consider removing 'async' from the method declaration or adding a 'yield' statement.

    static async System.Collections.Generic.IAsyncEnumerable<object> TestAsyncEnumerable() // error CS8420
    {
        throw new System.NotImplementedException();
    }

To get around it, I'm throwing in a useless `yield break` in the body or omit `async` from the declaration.

    static async System.Collections.Generic.IAsyncEnumerable<object> TestAsyncEnumerable() // ok
    {
        throw new System.NotImplementedException();
        yield break;
    }
    static System.Collections.Generic.IAsyncEnumerable<object> TestAsyncEnumerable() // also ok
    {
        throw new System.NotImplementedException();
    }

The compiler allows this fine for synchronous generators (which I understand is treated as a regular method that returns an enumerable).

    static System.Collections.Generic.IEnumerable<object> TestSyncEnumerable() // ok
    {
        throw new System.NotImplementedException();
    }

It does warn for regular asynchronous methods with CS1998.

> warning CS1998: This async method lacks 'await' operators and will run synchronously. Consider using the 'await' operator to await non-blocking API calls, or 'await Task.Run(...)' to do CPU-bound work on a background thread.

    static async System.Threading.Tasks.Task<object> TestAsync() // warning CS1998
    {
        throw new System.NotImplementedException();
    }

I think it should be allowed and not error for the async generator case.

**Comment by HaloFour** [comment, marked as answer]
You can omit the `async` keyword until you're ready to implement those methods.
